### DISCUSSION TITLE
169 INFO: UPX is available but is disabled on non-Windows due to known compatibility problems.

### DISCUSSION THREAD

**Discussion by OstapTelychko** [pyinstaller/pyinstaller#8922] [Help]
I have a question regarding UPX on Linux. I understand the meaning of this warning, but I’ve seen how it optimizes the output folder size on Windows. For instance, in my case, it reduced the size from 109 MB to 70 MB.

Now, I want to know if I can safely run UPX on some Linux binaries without breaking compatibility. Is it possible that all binaries could become incompatible after compression? I’m unsure.

Currently, my output folder size is 144 MB on Linux, and I’m looking to use UPX to reduce the update time for my users.

**System Information:**
__OS__: Linux Mint 22 (Wilma)
__PyInstaller Version__: 6.11.1
__Python Version__: 3.11.4
__Modules Used__: PySide6, SQLAlchemy, sqlite3, Alembic, threading, unittest

**Research So Far**:
I have searched all open and closed issues, as well as discussions, but I couldn’t find any mention of UPX. Perhaps I missed something, but I did my best to avoid creating duplicate issues or discussions.
If you know of any existing UPX-related issues or discussions that address my concern, please share them here. I will close this discussion if it is redundant

**Comment by bwoodsend** [comment]
Easiest way to explain is just to try running UPX on Linux. E.g. on latest Ubuntu:
```console
$ cat > test.c
#include <stdio.h>
void main() {
  printf("hello\n");
}
$ gcc test.c
$ upx a.out
$ ./a.out 
Segmentation fault
```
Try running UPX on your own project.

**Comment by jrd** [comment]
Work for me.

**Comment by OstapTelychko** [comment]
if anyone is intrested that's an update on may 18 2025
Summary: UPX 5.0.1 Compatibility with PyInstaller on Linux

UPX 5.0.1 brings improvements but still faces important limitations when used with PyInstaller on Linux, especially related to ELF internals.
**Where UPX 5.0.1 works:**
  Compressing standalone Linux executables (ELF binaries with simple layouts) — especially those with only two PT_LOAD segments (R+E for code and R+W for data).

  Works on most glibc-based distributions (e.g., Ubuntu, Debian, Fedora).

  Can reduce binary size significantly for PyInstaller one-folder mode, if .so files are excluded from compression.

**Where UPX 5.0.1 fails or breaks:**
  **Shared libraries (.so files):**
      UPX cannot reliably compress shared libraries on Linux.
      The dynamic linker (ld.so) does not execute UPX’s decompression stub.
      Attempting to compress .so files often leads to segmentation faults or runtime errors.

  **ELF binaries with more than 2 PT_LOAD segments:**
      Many modern binaries (especially PySide6 or Qt-based apps) include 3–4 segments (e.g., for RELRO, TLS, or extra RO data).
      UPX historically only supports 2 segments and will reject or corrupt binaries with layouts it doesn’t fully support.
        
  **PyInstaller --onefile mode:**
      The self-extracting stub created by PyInstaller --onefile is complex and often breaks after UPX compression.
      Segfaults, unpack failures, or runtime hangs are common.

**My recommendation**: Use UPX only on standalone executables, not .so files. Avoid using UPX with PyInstaller `--onefile` on Linux. Always test the compressed binary on the target platform, especially if using Qt, PySide6, or dynamically loaded libraries.

**Reply by rokm** [reply]
> Can reduce binary size significantly for PyInstaller one-folder mode, if .so files are excluded from compression.

If .so files are excluded from compression, what is there left to compress to significantly reduce binary size?

The bootloader part of main executable is ~55 kB uncompressed. The rest of the size of main executable comes from the embedded PKG archive (and PYZ archive embedded in that), and UPX touching any of that is likely to break the application.

There might be other collected 3rd party helper executables, but the most common ones (such as QtWebEngineProcess from PySide/PyQt) are unlikely to be successfully compressed, as per your own notes...

**Reply by OstapTelychko** [reply]
I have to admit UPX in current state is quite useless for pyinstaller since as you mentioned most of libraries are shared. But some *.so files will be compressed, you just have to verify if they still work after compression).
